go-controller/pkg/metrics/master.go

```diff
@@ -477,99 +507,239 @@ const (
+	// Override command line to disable this functionality if cluster node size is large to protect southbound database
+	if count, err := getNodeCount(kube); err != nil && count > nodeLimit {
+		klog.Infof(stopMsg, nodeLimit)
+		return
 	}
-}
+	// create a config duration stop channel because we want to coordinate stopping of this func if stop sig received
+	// or node count grows over a limit
+	cdStop := make(chan struct{})
+	// stop sig or node count can close cdStop
+	cpr.runStopSignal(kube, cdStop, stop)
+	cpr.runAvailableUpdate(kube, cdStop)
+	// create a map with key as the name of the k8 kind you want to measure and value with the histogram to report
+	// the duration to configure the kind to all nodes. Suggestion to name the key like so: {k8 kind}. E.g. pod
+	cpr.reportKind = make(map[string]prometheus.Histogram)
+	// register metrics which will take account of configuration start time and the time it
+	// takes OVN to configure all nodes. Note that the result will only give an upper bound limit of how long it
+	// takes to roll out the configuration to all nodes. To make matters more uncertain, the measurements are currently
+	// sampled.
+	cpr.reportKind[PodKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_pod_duration_seconds",
+		Help:      "Upperbound time to configure a pod from when its seen after pod handler and config is applied to all nodes",
+		Buckets:   prometheus.ExponentialBuckets(.01, 3, 11),
+	})
+	prometheus.MustRegister(cpr.reportKind[PodKindName])
+	cpr.reportKind[ServiceKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_service_duration_seconds",
+		Help:      "Upperbound time to configure a service from when its popped off a work queue and config is applied to all nodes",
+		Buckets:   prometheus.ExponentialBuckets(.01, 3, 11),
+	})
+	prometheus.MustRegister(cpr.reportKind[ServiceKindName])
+
+	prometheus.MustRegister(metricConfigDurationStartCount)
+	prometheus.MustRegister(metricConfigDurationEndCount)
+	prometheus.MustRegister(metricConfigDurationOvnLatency)
+
+	cpr.enabled = true
+
+	nbClient.Cache().AddEventHandler(&cache.EventHandlerFuncs{
+		UpdateFunc: func(table string, old model.Model, new model.Model) {
+			if table != nbGlobalTable {
+				return
+			}
+			// disable adding items to queue if functionality is disabled due to large node count
+			if !cpr.enabled {
+				return
+			}
 
-func (cpr *ControlPlaneRecorder) addPod(podUID kapimtypes.UID, t time.Time) {
-	cpr.podRecords[podUID] = &record{timestamp: t, timestampType: firstSeen}
+			oldRow := old.(*nbdb.NBGlobal)
+			newRow := new.(*nbdb.NBGlobal)
+
+			if oldRow.HvCfg != newRow.HvCfg && oldRow.HvCfgTimestamp != newRow.HvCfgTimestamp && newRow.HvCfgTimestamp > 0 {
+				if !cpr.queueFull() {
```

Comment:
I think we should not ignore events from the databases (both nb and sb) even if the queue is full, because it can break the chain of events, e.g. missing LSP event will lead to all the rest of pod events being ignored because of the wrong `TimestampType`, and for config duration it can lead to the wrong duration being recorded if `hv_cfg` update was missed for the previous record
But we still can use it for adding-record events like AddPod or RecordConfigDuration

Reply:
Makes sense. Will do.